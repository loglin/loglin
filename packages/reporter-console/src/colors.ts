import type { LogLevels } from 'loglin'

export interface LevelColor {
  text: string
  labelBackground: string
  labelText: string
}

export type LevelColors = Record<LogLevels, LevelColor>
export type ColorsConfig = Partial<Record<LogLevels, Partial<LevelColor>>>

const labelTextColor = '#384A57'
const THEME_TEXT_COLOR = 'THEME_TEXT_COLOR'
const textColor = THEME_TEXT_COLOR

export const isThemeTextColor = (color: string) => color === THEME_TEXT_COLOR

export const defaultColors: LevelColors = {
  fatal: {
    text: textColor,
    labelText: labelTextColor,
    labelBackground: '#F54761',
  },
  error: {
    text: textColor,
    labelText: labelTextColor,
    labelBackground: '#F54761',
  },
  warn: {
    text: textColor,
    labelText: labelTextColor,
    labelBackground: '#FFC966',
  },
  log: {
    text: textColor,
    labelText: labelTextColor,
    labelBackground: '#FFF',
  },
  info: {
    text: textColor,
    labelText: labelTextColor,
    labelBackground: '#51D1FB',
  },
  success: {
    text: textColor,
    labelText: labelTextColor,
    labelBackground: '#05E17A',
  },
  debug: {
    text: textColor,
    labelText: labelTextColor,
    labelBackground: '#AAA',
  },
}

export function mergeColors(colors: ColorsConfig) {
  const mergedColors: ColorsConfig = {}

  for (const level in colors) {
    const defaultColorConfig = defaultColors[level as LogLevels]
    const colorConfig = colors[level as LogLevels]

    mergedColors[level as LogLevels] = {
      ...defaultColorConfig,
      ...colorConfig,
    }
  }

  return {
    ...defaultColors,
    ...mergedColors,
  } as LevelColors
}
