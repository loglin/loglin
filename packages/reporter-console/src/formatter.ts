import { createFormatter, LogLevels, FormatterFunction } from 'loglin'
import { bold } from 'chalk'
import {
  LevelColors,
  isThemeTextColor,
  mergeColors,
  defaultColors,
  ColorsConfig,
} from './colors'
import { chalkBgColor, chalkColor, formatError } from './utils'

export const createLabel = (
  level: LogLevels,
  colors: LevelColors = defaultColors
) => {
  const colorsConfig = colors[level]

  return chalkBgColor(colorsConfig.labelBackground)(
    `${bold(` ${chalkColor(colorsConfig.labelText)(level.toUpperCase())} `)}`
  )
}

// @TODO Remove explicit type
// https://github.com/microsoft/rushstack/issues/1050
export const basicFormatter: FormatterFunction = createFormatter(
  () => ({ level, message }) => `[${level.toUpperCase()}] ${message}`
)

// @TODO Remove explicit type
// https://github.com/microsoft/rushstack/issues/1050

export interface FancyFormatterOptions {
  colors?: ColorsConfig
}

export const fancyFormatter: FormatterFunction<FancyFormatterOptions> = createFormatter<FancyFormatterOptions>(
  (options) => {
    const colors = options?.colors ? mergeColors(options.colors) : defaultColors
    const labels: Partial<Record<LogLevels, string>> = {}

    return ({ level, message }) => {
      const textColor = colors[level].text

      let fancyMessage = isThemeTextColor(textColor)
        ? message
        : chalkColor(textColor)(message)

      if (message instanceof Error) {
        fancyMessage += formatError(message)
      }

      const cachedLabel = labels[level]
      if (cachedLabel) return `${cachedLabel} ${fancyMessage}`

      const label = createLabel(level, colors)
      labels[level] = label

      return `${label} ${fancyMessage}`
    }
  }
)
