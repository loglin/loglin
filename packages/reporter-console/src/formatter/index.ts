export * from './basic'
export * from './fancy'
