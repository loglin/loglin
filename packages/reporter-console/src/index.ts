export * from './reporter'
export * from './formatter'
