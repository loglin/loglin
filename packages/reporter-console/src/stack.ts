import { sep } from 'path'

export function parseStack(stack: string) {
  const cwd = process.cwd() + sep

  return stack
    .split('\n')
    .splice(1)
    .map((l) => l.trim().replace('file://', '').replace(cwd, ''))
}

export function formatStack(error: Error) {
  if (!error.stack) return
  return ['\n', ...parseStack(error.stack).map((line) => '  ' + line)]
}
