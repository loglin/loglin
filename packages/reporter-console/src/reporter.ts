import {
  createReporter,
  isFiltered,
  getLevelNumber,
  ReporterFunction,
} from 'loglin'
import { basicFormatter } from './formatter'

// @TODO Remove explicit type
// https://github.com/microsoft/rushstack/issues/1050
export const consoleReporter: ReporterFunction = createReporter((options) => {
  const formatter = options?.formatter || basicFormatter()
  const filters = options?.filters

  return (info) => {
    if (filters && filters.length && isFiltered(filters, info)) return false

    const formattedMessage = formatter(info)
    const logger =
      getLevelNumber(info.level) < 2 ? process.stderr : process.stdout

    logger.write(formattedMessage + '\n')
  }
})
