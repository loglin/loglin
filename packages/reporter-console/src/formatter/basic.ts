import type { Formatter } from 'loglin'
import { formatStack } from '../stack'

export function basicFormatter(): Formatter {
  return ({ level, message }) => {
    const formattedMessage =
      message instanceof Error
        ? `${message.message}${formatStack(message)?.join('\n') + '\n'}`
        : message

    return `[${level.toUpperCase()}] ${formattedMessage}`
  }
}
