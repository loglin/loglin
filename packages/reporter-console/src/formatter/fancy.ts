import { Formatter, LogLevel } from 'loglin'
import {
  defaultColors,
  LevelColors,
  ColorsConfig,
  mergeColors,
  chalkBgColor,
  chalkColor
} from '../colors'
import { formatStack } from '../stack'

export function fancyFormatter(options?: { colors: ColorsConfig }): Formatter {
  const colors = options?.colors ? mergeColors(options.colors) : defaultColors
  const labels: Partial<Record<LogLevel, string>> = {}

  return ({ level, message }) => {
    const textColor = colors[level].text
    let fancyMessage = message

    if (!labels[level]) {
      labels[level] = createLabel(level, colors)
    }

    if (message instanceof Error) {
      fancyMessage =
        (textColor ? chalkColor(textColor)(message.message) : message.message) +
        colorStack(message, level, colors)
    } else if (textColor) {
      fancyMessage = chalkColor(textColor)(message)
    }

    return `${labels[level]} ${fancyMessage}`
  }
}

export function createLabel(level: LogLevel, colors: LevelColors) {
  const colorsConfig = colors[level]

  return chalkBgColor(colorsConfig.labelBackground)(
    chalkColor(colorsConfig.labelText).bold(` ${level.toUpperCase()} `)
  )
}

export function colorStack(error: Error, level: LogLevel, colors: LevelColors) {
  const stack = formatStack(error)
  if (!stack) return

  return (
    stack
      .map((line) =>
        line
          .replace(/^  at +/, (m) => chalkColor(colors[level].stackAtColor)(m))
          .replace(
            /\((.+)\)/,
            (_, m) => `(${chalkColor(colors[level].stackFileColor)(m)})`
          )
      )
      .join('\n') + '\n'
  )
}
