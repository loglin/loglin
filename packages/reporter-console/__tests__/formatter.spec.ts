import { LogLevels } from 'loglin'
import { createLabel, basicFormatter, fancyFormatter } from '../src/formatter'
import { mergeColors } from '../src/colors'

describe('createLabel', () => {
  test('should create label with default colors', () => {
    expect(createLabel(LogLevels.Success)).toMatchSnapshot()
  })

  test('should create label with customized colors', () => {
    const colors = mergeColors({ success: { labelBackground: '#FFC7D4' } })
    expect(createLabel(LogLevels.Success, colors)).toMatchSnapshot()
  })
})

describe('basicFormatter', () => {
  test('should format message', () => {
    const formatter = basicFormatter()
    expect(
      formatter({ level: LogLevels.Info, message: 'Info' })
    ).toMatchSnapshot()
  })
})

describe('fancyFormatter', () => {
  const getTestError = () => {
    const error = new Error('Something went wrong')
    error.stack =
      'Error\n' + ['at line 1 (path)', 'at line 2 (path)'].join('\n')

    return error
  }

  test('should format message with default colors', () => {
    const formatter = fancyFormatter()
    expect(
      formatter({ level: LogLevels.Info, message: 'Info' })
    ).toMatchSnapshot()
  })

  test('should format message with custom colors', () => {
    const formatter = fancyFormatter({
      colors: {
        info: {
          labelText: '#FFF',
          text: '#00FF00',
        },
      },
    })

    expect(
      formatter({ level: LogLevels.Info, message: 'Info' })
    ).toMatchSnapshot()
  })

  test('should format message if instanceof Error', () => {
    const formatter = fancyFormatter()
    const error = getTestError()

    expect(
      formatter({
        level: LogLevels.Error,
        message: error,
      })
    ).toMatchSnapshot()
  })

  test('should be colored message if instanceof Error and not theme color', () => {
    const formatter = fancyFormatter({
      colors: {
        error: {
          text: '#ff0000',
        },
      },
    })

    const error = getTestError()

    expect(
      formatter({
        level: LogLevels.Error,
        message: error,
      })
    ).toMatchSnapshot()
  })

  // @TODO How to test that correctly?
  test('should use cached label', () => {
    const formatter = fancyFormatter()

    for (let i = 0; i < 2; i++) {
      expect(
        formatter({ level: LogLevels.Success, message: 'Success' })
      ).toMatchSnapshot()
    }
  })
})
