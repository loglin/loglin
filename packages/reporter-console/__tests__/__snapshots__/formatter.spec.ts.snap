// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basicFormatter should format message 1`] = `"[INFO] Info"`;

exports[`createLabel should create label with customized colors 1`] = `" SUCCESS "`;

exports[`createLabel should create label with default colors 1`] = `" SUCCESS "`;

exports[`fancyFormatter should format message if instanceof Error 1`] = `
" ERROR  Error: Something went wrong

  at Object.<anonymous> (packages\\\\reporter-console\\\\__tests__\\\\formatter.spec.ts:54:18)
  at Object.asyncJestTest (node_modules\\\\jest-jasmine2\\\\build\\\\jasmineAsyncInstall.js:106:37)
  at node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:45:12
  at new Promise (<anonymous>)
  at mapper (node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:28:19)
  at node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:75:41
  at processTicksAndRejections (internal/process/task_queues.js:93:5)"
`;

exports[`fancyFormatter should format message with custom colors 1`] = `" INFO  Info"`;

exports[`fancyFormatter should format message with default colors 1`] = `" INFO  Info"`;

exports[`fancyFormatter should use cached label 1`] = `" SUCCESS  Success"`;

exports[`fancyFormatter should use cached label 2`] = `" SUCCESS  Success"`;
